### PR TITLE
mitchell/New virtual signup flow

### PR DESCRIPTION
**Pull request by mitchell-deriv**

### CODE REVIEW

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 1]
```diff
@@ -1,45 +1,53 @@
...
+        is_virtual,
+        upgrade_info;
+
+    const init = () => {
+        upgrade_info      = Client.getUpgradeInfo();
+        is_virtual        = Client.get('is_virtual');
+        el_welcome_container  = getElementById('welcome_container');
+    };
+
+    const getCanUpgrade = (upgrade_type ,  { can_upgrade_to } = upgrade_info) => can_upgrade_to.includes(upgrade_type);
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/welcome_page.js:19]: (throughout) can we use `ClientBase.getBasicUpgradeInfo` method instead of this?

Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:20]: done

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        el_welcome_container,
+        upgrade_info,
+        is_uk,
+        is_unwelcome_uk;
+
+    const init = () => {
+        is_virtual            = Client.get('is_virtual');
+        upgrade_info          = ClientBase.getBasicUpgradeInfo();
+        el_welcome_container  = getElementById('welcome_container');
+        is_uk                 = Boolean(Client.get('residence') === 'gb');
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/digital_options.js:21]: I believe `===` is already a boolean converter and it doesn't need another `Boolean` wrapper around it, no?

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        is_unwelcome_uk       = State.getResponse('get_account_status.status').some(status => status === 'unwelcome') && (/gb/.test(Client.get('residence')));
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/digital_options.js:22]: if we store `Client.get('residence') === 'gb'` in a variable we can use it in both of these.

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    };
+
+    const getCanUpgrade = (upgrade_type ,  { can_upgrade_to } = upgrade_info) => can_upgrade_to.includes(upgrade_type);
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/digital_options.js:25]:
```suggestion
    const getCanUpgrade = (upgrade_type, { can_upgrade_to } = upgrade_info) => can_upgrade_to.includes(upgrade_type);
```

File: src/javascript/app/pages/user/new_account/digital_options.js (hunk 3 above)
Comment (negarn) [reply src/javascript/app/pages/user/new_account/digital_options.js:22]: instead of .some can we also use .includes for the unwelcome?

Comment (negarn) [reply src/javascript/app/pages/user/new_account/digital_options.js:22]: also if this function is used in a few files, might be worth making it a shared function in client.js or similar

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    const onLoad = () => {
+        BinarySocket.wait('authorize', 'landing_company', 'get_settings', 'get_account_status').then(() => {
+            init();
+            if (Client.hasAccountType('real')) {
+                window.location.href = Client.defaultRedirectUrl();
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/digital_options.js:31]: we can't use pjax redirect?

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ 
+            getElementById('default').addEventListener('click', () => {
+                window.location.href = Client.defaultRedirectUrl();
+            });
+            
+            getElementById('financial').addEventListener('click', () => {
+                if (is_virtual && upgrade_info.can_upgrade_to.length){
+                    if (getCanUpgrade('svg')){window.location.href = Url.urlFor('/user/metatrader'); return;}
+                    if (getCanUpgrade('maltainvest')){window.location.href = Client.defaultRedirectUrl(); return;}
+                    if (getCanUpgrade('iom') && (is_uk && is_unwelcome_uk))window.location.href = Url.urlFor('/user/metatrader');
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/digital_options.js:43]: please fix the spacing and square brackets of the above three lines.

Comment (negarn) [reply src/javascript/app/pages/user/new_account/digital_options.js:43]: same for these, we can't use pjax load?

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                }
+            });
+        });
+    };
+
+    return {
+        onLoad,
+    };
+})();
+module.exports = DigitalOptions;
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/digital_options.js:53]:
```suggestion

module.exports = DigitalOptions;
```

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 8]
```diff
@@ -1,45 +1,59 @@
...
+        upgrade_info,
+        is_uk,
+        is_unwelcome_uk;
+
+    const init = () => {
+        upgrade_info          = ClientBase.getBasicUpgradeInfo();
+        is_virtual            = Client.get('is_virtual');
+        el_welcome_container  = getElementById('welcome_container');
+        is_uk                 = Boolean(Client.get('residence') === 'gb');
+        is_unwelcome_uk       = State.getResponse('get_account_status.status').some(status => status === 'unwelcome') && (/gb/.test(Client.get('residence')));
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:22]: same comments as the file above for these two.

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 9]
```diff
@@ -1,45 +1,59 @@
...
+    };
+    
+    const getCanUpgrade = (upgrade_type ,  { can_upgrade_to } = upgrade_info) => can_upgrade_to.includes(upgrade_type);
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:25]:
```suggestion
    const getCanUpgrade = (upgrade_type, { can_upgrade_to } = upgrade_info) => can_upgrade_to.includes(upgrade_type);
```

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 10]
```diff
@@ -1,45 +1,59 @@
...
+            init();
             if (Client.hasAccountType('real')) {
                 window.location.href = Client.defaultRedirectUrl();
                 showLoadingImage(el_welcome_container, 'dark');
             }
 
-            const upgrade_info = Client.getUpgradeInfo();
-            const el_upgrade_title = getElementById('upgrade_title');
-            let upgrade_btn_txt = '';
+            getElementById('default').addEventListener('click', () => {
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:35]: why does something have the generic id `default` 🤔 that sounds like a really bad id to have 😄

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 11]
```diff
@@ -1,45 +1,59 @@
...
+                /* eslint no-unused-expressions: ["error", { "allowTernary": true }] */
+                (getCanUpgrade('iom') || (is_uk && is_unwelcome_uk))
+                    ? window.location.href = Url.urlFor('/new_account/realws')
+                    : window.location.href = Client.defaultRedirectUrl();
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:39]: pjax redirect?

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 12]
```diff
@@ -1,45 +1,59 @@
...
-            if (upgrade_info.can_upgrade) {
-                const upgrade_btn = getElementById('upgrade_btn');
-                if (upgrade_btn) {
-                    upgrade_btn.html(createElement('span', { text: localize('Upgrade now') })).setAttribute('href', Url.urlFor(upgrade_url));
-                    upgrade_btn.classList.remove('button-disabled');
+            getElementById('cfd').addEventListener('click', () => {
+                if (is_virtual && upgrade_info.can_upgrade_to.length) {
+                    if (getCanUpgrade('svg')){window.location.href = Url.urlFor('/user/metatrader'); return;}
+                    if (getCanUpgrade('maltainvest')){window.location.href = Client.defaultRedirectUrl(); return;}
+                    if (getCanUpgrade('iom') && (is_uk && is_unwelcome_uk))window.location.href = Url.urlFor('/user/metatrader');
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:46]: spacing and pjax

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 13]
```diff
@@ -1,45 +1,59 @@
...
                 }
-            }
+            });
+
+            getElementById('d_ptions').addEventListener('click', () => {
+                if (is_virtual && upgrade_info.can_upgrade_to.length){
+                    if (getCanUpgrade('svg'))window.location.href = Client.defaultRedirectUrl();
+                    if (getCanUpgrade('maltainvest'))window.location.href = Url.urlFor('new_account/digital_options');
+                    if (getCanUpgrade('iom') && (is_uk && is_unwelcome_uk))window.location.href = Url.urlFor('/new_account/realws');
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:54]: same here

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 14]
```diff
@@ -1,45 +1,59 @@
...
+                }
+            });
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:56]: where are we removing all the event listeners? they should be removed on onunload function or before the add event

File: src/javascript/app/pages/user/new_account/digital_options.js (hunk 2 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/digital_options.js:21]: Done

File: src/javascript/app/pages/user/new_account/digital_options.js (hunk 3 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/digital_options.js:22]: done all apart from .some change. It doesn't work with includes.

File: src/javascript/app/pages/user/new_account/digital_options.js (hunk 5 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/digital_options.js:31]: done

File: src/javascript/app/pages/user/new_account/digital_options.js (hunk 6 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/digital_options.js:43]: doen both

File: src/javascript/app/pages/user/new_account/digital_options.js (hunk 7 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/digital_options.js:53]: done

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 8 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:22]: done did new function in client base

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 9 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:25]: fixed

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 11 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:39]: done

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 12 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:46]: done

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 13 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:54]: done

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 14 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:56]: done also

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 15]
```diff
@@ -1,50 +1,99 @@
...
-                upgrade_btn_txt = upgrade_info.type[0] === 'financial'
-                    ? localize('Financial Account')
-                    : localize('Real Account');
+            d_options.addEventListener('click', onDOptions);
+        });
+    };
+
+    const onNotSure = () => {
+        /* eslint no-unused-expressions: ["error", { "allowTernary": true }] */
+        (getCanUpgrade('iom') || (is_uk && is_unwelcome_uk))
```
Comment (negarn) [inline src/javascript/app/pages/user/new_account/welcome_page.js:51]: why not write a normal if condition here so we don't need to disable eslint?

File: src/templates/app/new_account/welcome_page.jsx [hunk 16]
```diff
@@ -1,46 +1,72 @@
...
-        <span className='account-created' />
+const WelcomePage = () => {
 
-        <SeparatorLine className='gr-padding-5' invisible />
+    const dropdown = 'images/pages/welcome/ddown.svg';
+    const options = [
+        {
+            icon_list: ['images/pages/welcome/cfd.svg'],
+            title    : it.L('CFD'),
+            desc     : it.L('<strong>Maximise returns</strong> by risking more.'),
```
Comment (negarn) [inline src/templates/app/new_account/welcome_page.jsx:11]: <strong> should be a placeholder so translators don't change it

File: src/templates/app/new_account/welcome_page.jsx [hunk 17]
```diff
@@ -1,46 +1,72 @@
...
+            url      : it.url_for('user/metatrader'),
+            action_id: 'cfd',
+        },
+        {
+            icon_list: ['images/pages/welcome/doptions.svg'],
+            title    : it.L('Digital Options'),
+            desc     : it.L('Earn <strong>fixed returns</strong> by risking only what you put in'),
```
Comment (negarn) [inline src/templates/app/new_account/welcome_page.jsx:18]: same here

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 15 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:51]: In reality I dont know why eslint doesnt permit it. anyways ill change it:)

File: src/templates/app/new_account/welcome_page.jsx (hunk 16 above)
Comment (mitchell-deriv) [reply src/templates/app/new_account/welcome_page.jsx:11]: done

File: src/templates/app/new_account/welcome_page.jsx (hunk 17 above)
Comment (mitchell-deriv) [reply src/templates/app/new_account/welcome_page.jsx:18]: done

File: src/javascript/app/pages/user/new_account/welcome_page.js (hunk 15 above)
Comment (mitchell-deriv) [reply src/javascript/app/pages/user/new_account/welcome_page.js:51]: done

**Review by negarn — DISMISSED** [review]

File: scripts/config/pages.js [hunk 18]
```diff
@@ -21,6 +21,7 @@ module.exports = [
 
     ['trading',                                 'app/trade/trading',                           'default',    'SmartTrader'],
 
+    ['new_account/digital_options',             'app/new_account/digital_options',             'default',    'Digital Options'],
```
Comment (reza-fs) [inline scripts/config/pages.js:24]: (question) this `new_account` page is is reserved for Real Account creation, not sure `digital_options` belong here. Would you please elaborate?

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+        financial;
+
+    const init = () => {
+        is_virtual            = Client.get('is_virtual');
+        upgrade_info          = ClientBase.getBasicUpgradeInfo();
+        el_welcome_container  = getElementById('welcome_container');
+        is_uk                 = Client.get('residence') === 'gb';
+        is_unwelcome_uk       = Client.isUnwelcomeUk();
+        synthetics            = getElementById('default');
+        financial             = getElementById('financial');
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/digital_options.js:26]: We use `$` prefix as a reminder that this is a jQuery object throughout the code. 

For vanilla javascript objects, you can use `el_` prefix.

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+        upgrade_info,
+        is_uk,
+        is_unwelcome_uk,
+        synthetics,
+        financial;
+
+    const init = () => {
+        is_virtual            = Client.get('is_virtual');
+        upgrade_info          = ClientBase.getBasicUpgradeInfo();
+        el_welcome_container  = getElementById('welcome_container');
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/digital_options.js:22]:
```suggestion
        $welcome_container  = getElementById('welcome_container');
```

See below

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+    const onLoad = () => {
+        BinarySocket.wait('authorize', 'landing_company', 'get_settings', 'get_account_status').then(() => {
+            init();
+
+            if (Client.hasAccountType('real')) {
+                BinaryPjax.load(Client.defaultRedirectUrl());
+                showLoadingImage(el_welcome_container, 'dark');
+            }
+ 
+            synthetics.addEventListener('click', onSynthethics);
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/digital_options.js:40]: (nitpick) 
```suggestion
            synthetics.addEventListener('click', onClickSynthethics);
```

Same for financial event listener.

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+        if (is_virtual && upgrade_info.can_upgrade_to.length){
+            if (getCanUpgrade('svg')){
+                BinaryPjax.load(urlFor('/user/metatrader'));
+                return;
+            }
+            if (getCanUpgrade('maltainvest') || getCanUpgrade('malta')){
+                BinaryPjax.load(`${urlFor('trading')}?market=forex&formname=risefall`);
+                return;
+            }
+            if (getCanUpgrade('iom') && (is_uk && is_unwelcome_uk)) BinaryPjax.load(urlFor('/user/metatrader'));
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/digital_options.js:60]:
```suggestion
            if (getCanUpgrade('iom') && (is_unwelcome_uk)) BinaryPjax.load(urlFor('/user/metatrader'));
```

`is_unwelcome_uk` already covers the residence check.

File: src/javascript/app/pages/user/new_account/digital_options.js [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+        } else {
+            BinaryPjax.load(Client.defaultRedirectUrl());
+        }
+    };
+
+    const onUnload = () => {
+        synthetics.removeEventListener('click', onSynthethics);
+        financial.removeEventListener('click', onFinancials);
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/digital_options.js:68]: :+1:

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 24]
```diff
@@ -1,50 +1,102 @@
...
+    const init = () => {
+        upgrade_info          = ClientBase.getBasicUpgradeInfo();
+        is_virtual            = Client.get('is_virtual');
+        el_welcome_container  = getElementById('welcome_container');
+        is_uk                 = Client.get('residence') === 'gb';
+        is_au                 = Client.get('residence') === 'au';
+        is_unwelcome_uk       = Client.isUnwelcomeUk();
+        not_sure              = getElementById('default');
+        cfd                   = getElementById('cfd');
+        d_options             = getElementById('d_ptions');
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/welcome_page.js:30]: (prefix dom elements)

File: src/javascript/app/pages/user/new_account/welcome_page.js [hunk 25]
```diff
@@ -1,50 +1,102 @@
...
+                BinaryPjax.load(Client.defaultRedirectUrl());
                 showLoadingImage(el_welcome_container, 'dark');
             }
+            if (is_au) BinaryPjax.load(urlFor('/user/metatrader'));
+
+            not_sure.addEventListener('click', onNotSure);
+
+            cfd.addEventListener('click', onCFD);
+
+            d_options.addEventListener('click', onDOptions);
```
Comment (reza-fs) [inline src/javascript/app/pages/user/new_account/welcome_page.js:49]: (nitpick)
```suggestion
            not_sure.addEventListener('click', onClickNotSure);

            cfd.addEventListener('click', onClickCFD);

            d_options.addEventListener('click', onClickDigitalOptions);
```

**Pull request closed without merging**